orchagent/routeorch.cpp

```diff
@@ -997,8 +1154,94 @@ bool RouteOrch::addRoute(sai_object_id_t vrf_id, const IpPrefix &ipPrefix, const
-        if (status != SAI_STATUS_SUCCESS)
+        object_statuses.emplace_back();
+        sai_status_t status = gRouteBulker.create_entry(&object_statuses.back(), &route_entry, 1, &route_attr);
+        if (status == SAI_STATUS_ITEM_ALREADY_EXISTS)
+        {
+            SWSS_LOG_ERROR("Failed to create route %s with next hop(s) %s: already exists in bulker",
+                    ipPrefix.to_string().c_str(), nextHops.to_string().c_str());
+            return false;
+        }
+    }
+    else
+    {
+        /* Set the packet action to forward when there was no next hop (dropped) */
+        if (it_route->second.getSize() == 0)
+        {
+            route_attr.id = SAI_ROUTE_ENTRY_ATTR_PACKET_ACTION;
+            route_attr.value.s32 = SAI_PACKET_ACTION_FORWARD;
+
+            object_statuses.emplace_back();
+            gRouteBulker.set_entry_attribute(&object_statuses.back(), &route_entry, &route_attr);
+        }
+
+        route_attr.id = SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID;
+        route_attr.value.oid = next_hop_id;
+
+        /* Set the next hop ID to a new value */
+        object_statuses.emplace_back();
+        gRouteBulker.set_entry_attribute(&object_statuses.back(), &route_entry, &route_attr);
+    }
+    return false;
+}
+
+bool RouteOrch::addRoutePost(RouteBulkContext& ctx, sai_object_id_t vrf_id, const IpPrefix &ipPrefix, const NextHopGroupKey &nextHops)
+{
+    SWSS_LOG_ENTER();
+
+    auto& object_statuses = ctx.object_statuses;
+
+    if (object_statuses.empty())
+    {
+        // Something went wrong before router bulker, will retry
+        return false;
+    }
+
+    /* next_hop_id indicates the next hop id or next hop group id of this route */
+    sai_object_id_t next_hop_id;
+
+    /* The route is pointing to a next hop */
+    if (nextHops.getSize() == 1)
+    {
+        NextHopKey nexthop(nextHops.to_string());
+        if (nexthop.ip_address.isZero())
+        {
+            next_hop_id = m_intfsOrch->getRouterIntfsId(nexthop.alias);
+            /* rif is not created yet */
+            if (next_hop_id == SAI_NULL_OBJECT_ID)
+            {
+                SWSS_LOG_INFO("Failed to get next hop %s for %s",
+                        nextHops.to_string().c_str(), ipPrefix.to_string().c_str());
+                return false;
+            }
+        }
+        else
+        {
+            if (!m_neighOrch->hasNextHop(nexthop))
+            {
+                SWSS_LOG_INFO("Failed to get next hop %s for %s",
+                        nextHops.to_string().c_str(), ipPrefix.to_string().c_str());
+                return false;
+            }
+        }
+    }
+    /* The route is pointing to a next hop group */
+    else
+    {
+        if (!hasNextHopGroup(nextHops))
+        {
+            // Previous added an temporary route
+            auto& tmp_next_hop = ctx.tmp_next_hop;
+            addRoutePost(ctx, vrf_id, ipPrefix, tmp_next_hop);
+            return false;
```

Comment:
This statement seems redundant as it is recursive.

Reply:
Yes it is recursive.
But it is necessary. Remove it will change behavior.

---
In reply to: [410513353](https://github.com/Azure/sonic-swss/pull/1238#discussion_r410513353) [](ancestors = 410513353)